### PR TITLE
Provide resilience to CP connection for domain V2

### PR DESCRIPTION
**Pull request by leleueri**
This PR introduce caching to provide : 

* session caching for domain V1
* session caching & resilience to a unreachable CP for domain V2

fixes AM-3476

### CODE REVIEW

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/user/impl/UserStoreImpl.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class UserStoreImpl implements UserStore {
+
+    private Cache<String, User> idCache;
```
Comment (podlesrafal) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/user/impl/UserStoreImpl.java:39]: Make it final

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/user/impl/UserStoreImpl.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    private int ttlInSec;
```
Comment (podlesrafal) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/user/impl/UserStoreImpl.java:41]: make it final

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/user/impl/UserStoreImplV2.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+ * @author GraviteeSource Team
+ */
+public class UserStoreImplV2 implements UserStore {
+
+    private static final String SEPARATOR = ":";
+
+    private Cache<String, User> idCache;
+    private Cache<String, User> gisCache;
+
+    private int ttl;
```
Comment (podlesrafal) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/user/impl/UserStoreImplV2.java:45]: make all those fields final

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/user/impl/UserStoreImplV2.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    public Maybe<User> getByInternalSub(String gis) {
+        return gisCache.rxGet(gis);
+    }
+
+    @Override
+    public Completable clear() {
+        return gisCache.rxClear().andThen(idCache.rxClear());
+    }
+
+    private String generateInternalSubFrom(String src, String externalId) {
```
Comment (podlesrafal) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/user/impl/UserStoreImplV2.java:87]: Suggestion. This is done here and in UserServiceImplV2. Maybe you can extract this to some utils, or at least keep SEPARATOR variable in one place? Currently there are 3 places where SEPARATOR is defined.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/test/java/io/gravitee/am/gateway/handler/common/user/UserServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@RunWith(MockitoJUnitRunner.class)
```
Comment (podlesrafal) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/test/java/io/gravitee/am/gateway/handler/common/user/UserServiceTest.java:50]: Can you use Junit5 with @ExtendWith(MockitoExtension.class)?

**Review by podlesrafal — APPROVED** [review]

**Review by lgw-gravitee — APPROVED** [review]

**Pull request merged**